tests/test_helper/bats/ovsdb_schema_upgrade.bats

```diff
@@ -0,0 +1,150 @@
+# This is a bash shell fragment -*- bash -*-
+
+load "test_helper/setup_teardown/$(basename "${BATS_TEST_FILENAME//.bats/.bash}")"
+
+setup() {
+    load test_helper/lxd.bash
+    load test_helper/common.bash
+    load test_helper/microovn.bash
+    load test_helper/ovsdb.bash
+    load ../.bats/bats-support/load.bash
+    load ../.bats/bats-assert/load.bash
+
+    # Ensure TEST_CONTAINERS is populated, otherwise the tests below will
+    # provide false positive results.
+    assert [ -n "$TEST_CONTAINERS" ]
+}
+
+@test "Test OVSDB cluster schema upgrade" {
+    echo "# Checking if SB or NB database schema changed from MicroOVN rev. $MICROOVN_SNAP_REV" >&3
+    local original_nb=""
+    local original_sb=""
+    local target_nb=""
+    local target_sb=""
+    local probe_container=""
+    probe_container=$(awk '{print $1}' <<< "$TEST_CONTAINERS")
+
+    # Get currently running OVN schema versions
+    original_nb=$(get_current_ovsdb_schema_version "$probe_container" "nb")
+    original_sb=$(get_current_ovsdb_schema_version "$probe_container" "sb")
+    # Get schema versions that are included in the tested MicroOVN snap
+    target_nb=$(get_ovsdb_schema_version_from_snap "$probe_container" "$MICROOVN_SNAP_PATH" "nb")
+    target_sb=$(get_ovsdb_schema_version_from_snap "$probe_container" "$MICROOVN_SNAP_PATH" "sb")
+
+    local requires_upgrade_check=0
+
+    if [ "$original_nb" != "$target_nb" ] || [ "$original_sb" != "$target_sb" ]; then
+        ((++requires_upgrade_check))
+    fi
+
+    # Skip this test if there's no change between schemas in running MicroOVN version
+    # (installed from snap store) and tested MicroOVN version (installed from source)
+    if [ "$requires_upgrade_check" -eq 0 ]; then
+        skip "OVN database schemas did not change. Skipping this check"
+    fi
+
+    declare -a UPGRADED_CONTAINERS=()
+    read -r -a OLD_CONTAINERS <<< "$TEST_CONTAINERS"
+    local container_count="${#OLD_CONTAINERS[@]}"
+
+    # Based on  the version of original MicroOVN snap, it may not have necessary API to
+    # report expected schema version. In that case we'll expect an error message from these members.
+    # Otherwise we expect them to report old schema version
+    local old_member_msg_nb=""
+    local old_member_msg_sb=""
+    if [ "$MICROOVN_SNAP_REV" -le 395 ]; then
+        old_member_msg_nb="Missing API. MicroOVN needs upgrade"
+        old_member_msg_sb="Missing API. MicroOVN needs upgrade"
+    else
+        old_member_msg_nb="$original_nb"
+        old_member_msg_sb="$original_sb"
+    fi
+
+    for container in $TEST_CONTAINERS; do
```

Comment:
This loop iterates over all the containers and moves in a forward direction. This means that the current element is to be upgraded, any previous elements are already upgraded and any coming elements are not upgraded yet. As such we can infer all this information from the current array index.

Would it make sense to use that instead of keeping track of all these extra arrays? Something like:
```bash
read -r -a containers <<< "$TEST_CONTAINERS"
for ((i = 0; i < "${#containers[@]}"; i++)) {
    container=${containers[i]}
    echo "processing container $container"

    for ((j = 0; j < i; j++)) {
        echo "upgraded container ${containers[j]}"
    }

    for ((j = i + 1; j < "${#containers[@]}"; j++)) {
        echo "old container ${containers[j]}"
    }
    echo
}
```

Reply:
This looks reasonable +1